File: eng/docs/agent_skills/python_typing/examples/integration tests/assembly_wiring_test.py
from __future__ import annotations

from dataclasses import dataclass, field

import pytest

from dictionary.implementation.TL_CORE import ApiError, ErrorCategory, Result


@dataclass(frozen=True, slots=True)
class OrderAccepted:
    order_id: str
    status: str


@dataclass(frozen=True, slots=True)
class OrderGuardConfig:
    single_inflight: bool = True


@dataclass(slots=True)
class OrderGuardState:
    inflight_ids: set[str] = field(default_factory=set)


class EventCapture:
    def __init__(self) -> None:
        self.events: list[dict] = []

    def emit(self, event: dict) -> None:
        self.events.append(event)


class OrderProvider:
    def __init__(self) -> None:
        self.calls: list[str] = []

    def submit_order(self, order_id: str) -> Result[OrderAccepted]:
        self.calls.append(order_id)
        return Result(
            outcome="ok",
            ok=OrderAccepted(order_id=order_id, status="accepted"),
        )


class GuardedOrderPort:
    def __init__(self, inner: OrderProvider, cfg: OrderGuardConfig, obs: EventCapture) -> None:
        self.inner = inner
        self.cfg = cfg
        self.obs = obs
        self.state = OrderGuardState()

    def submit_order(self, order_id: str) -> Result[OrderAccepted]:
        if self.cfg.single_inflight and order_id in self.state.inflight_ids:
            self.obs.emit({"ev": "ORDER_REJECTED_DUPLICATE", "order_id": order_id})
            return Result(
                outcome="err",
                err=ApiError(
                    category=ErrorCategory.CONFLICT,
                    code="ORDER_ALREADY_IN_FLIGHT",
                    retryable=True,
                ),
            )

        self.state.inflight_ids.add(order_id)
        try:
            return self.inner.submit_order(order_id)
        finally:
            self.state.inflight_ids.discard(order_id)


def build_order_stack() -> GuardedOrderPort:
    cfg = OrderGuardConfig(single_inflight=True)
    obs = EventCapture()
    provider = OrderProvider()
    return GuardedOrderPort(inner=provider, cfg=cfg, obs=obs)


@pytest.mark.cmp_int
@pytest.mark.assembly_wiring_smoke
@pytest.mark.smoke
def test_bootstrap_builds_guarded_order_stack_and_minimal_call_succeeds() -> None:
    order_port = build_order_stack()

    result = order_port.submit_order("ORDER_001")

    assert isinstance(order_port, GuardedOrderPort)
    assert isinstance(order_port.inner, OrderProvider)
    assert order_port.cfg.single_inflight is True
    assert result == Result(
        outcome="ok",
        ok=OrderAccepted(order_id="ORDER_001", status="accepted"),
    )
    assert order_port.inner.calls == ["ORDER_001"]
    assert order_port.obs.events == []

File: eng/docs/agent_skills/csharp_typing/examples/integration_tests/AssemblyWiringTest.cs
// ============================================================
// Integration test: Assembly / wiring smoke
// Family: assembly_wiring_smoke
// Level: cmp_int (component integration — full stack assembly)
// ============================================================
//
// Verifies that the bootstrap/wiring builds a correctly
// assembled component stack and that a minimal call succeeds.

using FluentAssertions;
using Xunit;

// Adjust these usings to your real namespace.

namespace Acme.Orders.Tests.Integration;

// ============================================================
// Domain types (self-contained example)
// ============================================================

internal sealed record OrderAcceptedResult(string OrderId, string Status);

internal sealed record OrderGuardConfig(bool SingleInflight = true);

internal sealed class OrderGuardState
{
    private readonly HashSet<string> _inflightIds = [];
    private readonly object _lock = new();

    public bool TryAdd(string orderId)
    {
        lock (_lock) { return _inflightIds.Add(orderId); }
    }

    public void Remove(string orderId)
    {
        lock (_lock) { _inflightIds.Remove(orderId); }
    }
}

internal sealed class WiringEventCapture
{
    private readonly List<Dictionary<string, object>> _events = [];

    public void Emit(Dictionary<string, object> ev) => _events.Add(ev);

    public IReadOnlyList<Dictionary<string, object>> Events => _events;
}

// ============================================================
// Provider
// ============================================================

internal sealed class OrderProvider
{
    public List<string> Calls { get; } = [];

    public OrderAcceptedResult SubmitOrder(string orderId)
    {
        Calls.Add(orderId);
        return new OrderAcceptedResult(orderId, "accepted");
    }
}

// ============================================================
// Guard wrapping provider
// ============================================================

internal sealed class GuardedOrderStack
{
    private readonly OrderProvider _inner;
    private readonly OrderGuardConfig _config;
    private readonly WiringEventCapture _obs;
    private readonly OrderGuardState _state = new();

    public OrderProvider Inner => _inner;
    public OrderGuardConfig Config => _config;
    public WiringEventCapture Obs => _obs;

    public GuardedOrderStack(
        OrderProvider inner,
        OrderGuardConfig config,
        WiringEventCapture obs)
    {
        _inner = inner;
        _config = config;
        _obs = obs;
    }

    public OrderAcceptedResult SubmitOrder(string orderId)
    {
        if (_config.SingleInflight && !_state.TryAdd(orderId))
        {
            _obs.Emit(new() { { "ev", "ORDER_REJECTED_DUPLICATE" }, { "order_id", orderId } });
            throw new InvalidOperationException("ORDER_ALREADY_IN_FLIGHT");
        }

        try
        {
            return _inner.SubmitOrder(orderId);
        }
        finally
        {
            _state.Remove(orderId);
        }
    }
}

// ============================================================
// Bootstrap / factory function
// ============================================================

internal static class OrderStackFactory
{
    public static GuardedOrderStack Build()
    {
        var config = new OrderGuardConfig(SingleInflight: true);
        var obs = new WiringEventCapture();
        var provider = new OrderProvider();
        return new GuardedOrderStack(provider, config, obs);
    }
}

// ============================================================
// Tests
// ============================================================

[Trait("Level", "cmp_int")]
[Trait("Family", "assembly_wiring_smoke")]
public sealed class AssemblyWiringTest
{
    [Fact]
    public void BootstrapBuildsGuardedOrderStackAndMinimalCallSucceeds()
    {
        // Arrange — build via factory
        var orderPort = OrderStackFactory.Build();

        // Act
        var result = orderPort.SubmitOrder("ORDER_001");

        // Assert — correct types assembled
        orderPort.Should().BeOfType<GuardedOrderStack>();
        orderPort.Inner.Should().BeOfType<OrderProvider>();
        orderPort.Config.SingleInflight.Should().BeTrue();

        // Assert — call succeeded
        result.Should().Be(
            new OrderAcceptedResult("ORDER_001", "accepted"));

        // Assert — provider was called, no guard events
        orderPort.Inner.Calls.Should().BeEquivalentTo(["ORDER_001"]);
        orderPort.Obs.Events.Should().BeEmpty();
    }
}
